feat(Codegen): Add list nullability migration guide (#3306)
aws-amplify/docs: docs/cli/migration/list-nullability.md

---
title: Amplify Codegen Models - List and list components nullability
description: Generating Amplify models with Amplify CLI x.y.z
---

### **What is changing?** 

Amplify CLI x.y.z has updated the codegen process to correctly set the optionality for types in models.

_Schema example_

```
type ListStringContainer @model {
  stringList: [String!]!
  stringNullableList: [String!]
  nullableStringList: [String]!
  nullableStringNullableList: [String]
}

> ```

Reviewer: Nit: it would be easier to keep track of which thing you're talking about if you maintained parallelism between the case names. Genericizing the type to "Element" in the name also makes it a bit clearer what you're referring to:

requiredElementRequiredList: [String!]!
optionalElementOptionalList: [String]
requiredElementOptionalList: [String!]
optionalElementRequiredList: [String]!

- `stringList` - the list component and the list is required
- `stringNullableList` - the list component is required, the list is optional
- `nullableStringList` - the list component is optional, the list is required
- `nullableStringNullableList` - the list component and list is optional

> The list component in this example is a String type, however, this applies for other types as well such as Int, Bool, and embedded types that you define yourself.

### **Why are we introducing this change?** 

This is to align the optionality of the generated Swift models as closely as possible to the type defined in the schema.

### **Who is impacted?** 

<amplify-block-switcher>

<amplify-block name="iOS">

Developers building an iOS app with Amplify DataStore or Amplify API generates Swift models by running the command `amplify codegen models`. 

 _Previous generated Swift code_

```swift
public struct ListStringContainer: Model {
  public var stringList: [String]
  public var stringNullableList: [String]
  public var nullableStringList: [String]?
  public var nullableStringNullableList: [String]?
  ...
}
```

> _Current code generated Swift model_

Reviewer: ```suggestion
_Current code generated Swift code_
```

```swift
public struct ListStringContainer: Model {
  public var stringList: [String]
  public var stringNullableList: [String]?
  public var nullableStringList: [String?]
  public var nullableStringNullableList: [String?]?
  ...
}
```

The difference between the current and previous code:

- `stringList` - No changes
- `stringNullableList` - the list was required and is now optional.
- `nullableStringList` - the list component was required and is now optional. The list was optional and is now required
- `nullableStringNullableList` - the list component was required and is now optional. 

</amplify-block>

</amplify-block-switcher>

### **When do I have to upgrade?**

> This is behind a feature flag in Amplify CLI x.y.z and will be deprecated by [TODO: DATE]. Developers with existing apps should upgrade to the latest CLI, set the feature flag, and update their app code to account for the change in optionality of the types. Developers building a new app will automatically generate code with the latest changes and no action is required.

Reviewer: "update their app code or their schema (see recommendations following)"

### **Where do I make these changes?**

1. Update Amplify CLI to the latest version

```
amplify upgrade
```

2. The version should be at least x.y.z

```

> amplify --v // at least x.y.z

Reviewer: nit: shell comment delimiter is `#`, not `//`
```suggestion
amplify --v    # at least x.y.z
```

```

3. If building an existing app, toggle the feature flag to true in TODO: PATH

4. Run `amplify codegen models` to generate the latest models.

> 5. Open the App and make sure the app compiles. Depending on your schema, you may be in the following scenarios.

Author: re: where do developers make these changes? I think for iOS at least, their models are linked in the Xcode project, so it would be under the AmplifyModel folder. I can update this note in the iOS block below that they can view the latest generated models there, and the actually code changes are dependent on their app code

Reviewer 2: The generated models will be in `/amplify/generated/models` folder from the app root. I believe they are linked to AmplifyModel folder, so either paths work.

```
type ListStringContainer @model {
  requiredElementRequiredList: [String!]!
  requiredElementOptionalList: [String!]
  optionalElementRequiredList: [String]!
  optionalElementOptionalList: [String]
}
```

In this example, there are four fields with different combinations of optionality:

> - `requiredElementRequiredList` - the list component and the list is required
> - `requiredElementOptionalList` - the list component is required, the list is optional
> - `optionalElementRequiredList` - the list component is optional, the list is required
> - `optionalElementOptionalList` - the list component and list is optional

Reviewer: Does this clarify further?

```suggestion
- `requiredElementRequiredList` - the list itself is required. Elements it contains must be non-null. Empty lists can only be represented as an empty array.
- `requiredElementOptionalList` - the list itself is optional. If present, elements it contains must be non-null. Empty lists could be represented as either an empty array or a null field.
- `optionalElementRequiredList` - the list itself is required. If present, elements it contains may be null.
- `optionalElementOptionalList` - the list itself is optional. If present, elements it contains may be null.
```

```swift
public struct ListStringContainer: Model {
  public var requiredElementRequiredList: [String]
  public var requiredElementOptionalList: [String]

>   public var nullableStringList: [String]?

Reviewer: ```suggestion
  public var optionalElementRequiredList: [String]?
```

Author: that's odd, the current file already shows this change

  public var optionalElementOptionalList: [String]?
  ...
}
```

_Current code generated Swift code_

```swift
public struct ListStringContainer: Model {
  public var requiredElementRequiredList: [String]
  public var requiredElementOptionalList: [String]?

>   public var nullableStringList: [String?]

Reviewer: ```suggestion
  public var optionalElementRequiredList: [String?]
```

Author: same for this one
